Fix data race in InputHandler's

bhennion opened the pull request:
This PR should fix #3579 (and possibly #4038).

It is not as clean as I'd want, and I'd like some advice on how this kind of mutex handling is best implemented. 
At least, there does not seem to be any deadlock happening (but I only tested it briefly).

@talkdirty: would you mind checking this PR out and report on whether or not this fixes #4038?

Febbe reviewed (commented):
**Literature**

For Concurrency tips you should read [this](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#cpcon-concurrency)   
A book which is very good: Concurrency in Action from Anthony Williams

Important is: [CP.31](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-data-by-value) [CP.43](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-time) [CP.44](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-name) [CP.50](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rconc-mutex)

I think you acquire the lock for too long. Try to find as many blocks which have a minimum size.

Febbe commented on src/control/tools/InputHandler.cpp:
```diff
@@ -32,7 +31,7 @@ void InputHandler::resetShapeRecognizer() {
 }
 
-void InputHandler::createStroke(Point p) {
+void InputHandler::createStroke(Point p, const std::lock_guard<std::recursive_mutex>&) {
```
**General Todo for a refactoring:** I am not happy with the naming: createXXX should either return a Stroke or a unique / shared pointer to a Stroke. Not void.

Febbe replied:
this for example is a perfect way to show, that the locking is too wide. The only thing, which must be protected is the line 37: `stroke = new Stroke();`
also this line can be moved to the end of the function / the return value by replacing the line with `auto stroke = std::make_unique<Stroke>(); ...; this->stroke = stroke.release();`

Febbe commented on src/control/tools/InputHandler.cpp:
```diff
@@ -32,7 +31,7 @@ void InputHandler::resetShapeRecognizer() {
 }
 
-void InputHandler::createStroke(Point p) {
+void InputHandler::createStroke(Point p, const std::lock_guard<std::recursive_mutex>&) {
```
Don't provide lock_guard's in the function header, certainly never as const reference. It may be useful, to assure, that a lock is acquired. But then it has to be either replaced with std::unique_lock (value semantic or non-const reference).

bhennion commented:
I guess all this means we'll need a deeper PR, which may no longer be suitable for the release branch.
I'll try to integrate your feedback while keeping the changes to a minimum. We'll see if I can figure something out.

Febbe commented:
No, I think the easiest solution would be, to replace that pointer with a shared_ptr. And you just use a std::swap on it, when it is updated. But that means that it is not allowed to change the stroke without making a copy of it (and swap the new shared_ptr into it). Then you only have to protect the write operation on the shared_ptr (unless you are using atomic<shared_ptr>)

Febbe commented:
Also removing all those obsolete references to the lock_guard will reduce the diff.

bhennion commented:
I made a second attempt. Things are probably more like @Febbe recommended in StrokeHandler and BaseStrokeHandler. However, I do not see how to do things properly in SplineHandler without refactoring.
Basically, the mutex needs to protect all the data (the stroke, the knots, and the tangent vectors), and those are used (either read or write) on almost every line of code.

I think the best way to handle this would be to split the data into "old and fixed points" and "moving spline segment". The old and fixed data should be painted onto a mask in the main loop (the mask begin protected by a mutex, ofc). The moving segment should be shared by value with the drawing thread. Those kind of changes go beyond what can be done on the release branch, I think.

I'd say we have a choice to make for the SplineHandler: either we leave the data race be, or we use this "overlocking" version.

Febbe commented on src/control/tools/BaseStrokeHandler.cpp:
```diff
@@ -150,30 +139,31 @@ void BaseStrokeHandler::onButtonReleaseEvent(const PositionInputData& pos) {
     UndoRedoHandler* undo = control->getUndoRedoHandler();
 
-    undo->addUndoAction(std::make_unique<InsertUndoAction>(page, layer, stroke));
+    Stroke* s;
```
```suggestion
    std::unique_ptr<Stroke> s;
```

Febbe commented on src/control/tools/BaseStrokeHandler.cpp:
```diff
@@ -150,30 +139,31 @@ void BaseStrokeHandler::onButtonReleaseEvent(const PositionInputData& pos) {
+    Stroke* s;
+    {
+        std::lock_guard lock(strokeMutex);
+        s = stroke.release();
```
```suggestion
        s = std::move(stroke);
```

Febbe commented on src/control/tools/BaseStrokeHandler.h:
```diff
@@ -59,8 +61,11 @@ class BaseStrokeHandler: public InputHandler {
+    Point buttonDownPoint;  // used for tapSelect and filtering - never snapped to grid.
+    Point startPoint;       // May be snapped to grid
+
+    std::mutex strokeMutex;  // protects InputHandler::stroke
```
Why is this mutex not located next to the stroke?

bhennion replied:
Because BaseStrokeHandler::draw uses the stroke, so the stroke needs protecting in this derived class (of InputHandler). On the other hand StrokeHandler::draw uses a mask, so the stroke does not need protecting in this derived class.
I don't know how to make this cleanly...
